### ISSUE TITLE
Jobs get stuck after Redis reconnect 

### ISSUE THREAD

**Issue by royalpinto** [OptimalBits/bull#1873]
##  Description
When the Redis gets disconnected and connected again, the queue doesn't pick up any jobs. (Easy To Reproduce)

## Minimal, Working Test code to reproduce the issue.
```
const Queue = require('bull')
const Redis = require('ioredis')

const client = new Redis({
  host: 'localhost',
  port: 6379
})

const queue = new Queue('test', client)

client.on('ready', () => {
  console.log('client ready')
})

client.on('connect', () => {
  console.log('client connect')
})

client.on('close', () => {
  console.log('client close')
})

client.on('end', error => {
  console.log('client end', error)
})

client.on('reconnecting', error => {
  console.log('client reconnecting', error)
})

client.on('error', error => {
  console.log('client error', error.message)
})

queue
  .on('completed', (job, results) => {
    console.log('queue completed a job')
  })
  .on('error', async (error) => {
    console.error('queue error', error.message)
  })
  .on('failed', async (job, error) => {
    console.error('queue error', error.message)
  })

queue.process(async (job) => {
  console.log('running task')
})

setInterval(function () {
  queue.add({}).then(() => console.log('pushed task')).catch((e) => console.log('failed to push task'))
}, 10000)
```

## Bull version
Bull 3.18.0
Node 12 and 14
Redis 3.2.8

## Steps to reproduce.
- Run the above script
- Bull JS start accepting and processing Jobs
- Stop Redis
- Bull JS will keep retrying
- Start Redis
- Bull JS will connect in a few milliseconds
- At this stage, the queue can push the jobs and **but it won't process any jobs**. :-( 

## Things we tried
- Tried explicitly the reconnect option of `ioredis`. https://github.com/luin/ioredis#auto-reconnect and https://github.com/luin/ioredis#reconnect-on-error But issue don't seem to be coming from `ioredis`.

**Comment by manast** [comment]
I tested the code above (I just changed the timeout to be 1 second for faster testing), it is working perfectly on Mac os, I tried many times to stop the redis server and re-start it and it always continue processing as soon as the redis server is online. Is there anything else you can share in order to reproduce the issue?

**Comment by royalpinto** [comment]
@manast It's strange only on one instance I could reproduce, but other places I could not reproduce. I will try to investigate more on this.

However, there is a similar issue we found which is reproducible everywhere. The scenario is below:
- Stop Redis Server
- Start above snippet
- The snippet will keep trying to connect
- Start the Redis server
- The snippet will eventually connect in a few seconds.
=> Now **it cannot push or process any jobs**. It rather throws "UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:6379". (This is for the client but something similar for the subscriber too. Even if we push from another process, the subscriber won't pick up the jobs).

I tried analyzing the code and found that inside the `Queue` class the `this._initializing` promise is initialized only once in the constructor and it will be rejected because Redis will not be running during the startup. When it eventually connects, `this._initializing` will be the same old state so throws an error. It should rather re-initialize.

https://github.com/OptimalBits/bull/blob/develop/lib/queue.js#L148

**Comment by manast** [comment]
okok, so you mean if redis is not running when the bull starts, I will look into it.

**Comment by ruscon** [comment]
@royalpinto @manast yep, I have the same issue.

**Comment by mbieser** [comment]
+1

**Comment by manast** [comment]
I marked this as an enhancement because I consider it a new requirement, i.e. being able too work when starting without an available redis connection if that connection becomes available at a later time. May require some non trivial refactoring.

**Comment by geniousphp** [comment]
Actually, I can reproduce this scenario
- Run bull 
- Stop Redis
- Wait for a moment and start again Redis
- Bull will not process any tasks
- Restart bull, then Bull will work normally
I think this is a bug

**Comment by koiszzz** [comment]
I have met the same issue. when redis server came into crash for a long time, longer than redis reconnecting  times , the job just add into the queue and  just waiting,never start

**Comment by manast** [comment]
@koiszzz I cannot find how many times ioredis reconnects before it gives up, maybe it tries to reconnect forever or maybe not, this should be investigated, however I tested extensively reconnects in latest version of Bull and could not reproduce any issue. If anybody does please post some instructions on how to reproduce it, ideally a piece of code I can run so I can take a deeper look into it.

**Comment by koiszzz** [comment]
I post an bull test project [bull-test](https://github.com/koiszzz/bull-test), 
start redis server and then start the express server , browser the url : localhost:3000/add, the console will print job done.
then stop the redis server, and wait for a moment . 
restart the redis server, browser the url : localhost:3000/add, you will find the job add to the queue, but then state will be stuck in 'waiting'

**Comment by manast** [comment]
@koiszzz I tested your code. I guess you are testing using a redis instance inside docker? I experience a bug with ioredis reported here: https://github.com/luin/ioredis/issues/1285 where reconnect does not work if redis is running inside a docker container in MacOS at least. Your repo works well for me if I use a local redis instance but it does not work if I use a redis docker instance... So if this is your case I suggest you ping Luin to see if he can reproduce it, he has failed to reproduce it so far but if more people than me can reproduce it, then maybe it will get more attention... alternatively that I debug ioredis myself but not sure when I will find time for that...

**Comment by koiszzz** [comment]
Thanks for replying , the redis server is installed in docker. Maybe I should install a redis server without docker for test.

**Comment by givorra-ls** [comment]
Lol.... I had the same issue, I was reading your comments and doing some tests. I realised that when running redis via docker, the redis data has no persistence and is therefore deleted when the container is stopped and restarted... The "solution" is to start redis via docker-compose using a volume :)

**Comment by miwnwski** [comment]
I encountered this error and have been investigating, I'll share our findings here - maybe it helps bull queue. As far as I can understand, when the client reconnects, `Queue#processJobs` is not run. As far as I can tell, `Queue.prototype.process` calls `Queue#start` which calls `Queue#run` which calls `Queue#processJobs` which enters the queue in the forever loop state by throwing promises.

The problem is (as far as I can understand) that  `Queue#processJobs` calls `Queue#_processJobOnNextTick` on `nextTick` which can resolve if the Queue is closing or if an error is thrown.

This creates a state where the connection exists but the forever promise loop never executes.

Our purposed fix for this (at least from local - not tested in production)  is basically to call `queue.run(queue.concurrency)` on ready. Using the above example which I could use to reproduce the problem, this would be the fix:

```js
const Queue = require('bull')
const Redis = require('ioredis')

const client = new Redis({
  host: 'localhost',
  port: 6379
})

const queue = new Queue('test', client)
let reconnectingEvent = false 

client.on('ready', () => {
  console.log('client ready')
  if (reconnectingEvent) { // this is actually the fix, we just set a flag to not kick run unless we are restarting
    console.log('rerun queue run')
    reconnectingEvent = false
    queue.run(1)
  }
})

client.on('connect', () => {
  console.log('client connect')
})

client.on('close', () => {
  console.log('client close')
})

client.on('end', error => {
  console.log('client end', error)
})

client.on('reconnecting', error => {
  console.log('client reconnecting', error)
  reconnectingEvent = true
})

client.on('error', error => {
  console.log('client error', error.message)
})

queue
  .on('completed', (job, results) => {
    console.log('queue completed a job')
  })
  .on('error', async (error) => {
    console.error('queue error', error.message)
  })
  .on('failed', async (job, error) => {
    console.error('queue error', error.message)
  })

queue.process(async (job) => {
  console.log('running task')
})

setInterval(function () {
  queue.add({}).then(() => console.log('pushed task')).catch((e) => console.log('failed to push task'))
}, 10000)
```

I'm not sure if this is a "good" patch, but it works. Maybe maintainers have a better clue how to fix this cleanly (or describe to which I can PR a fix for).

**Comment by manast** [comment]
This sounds like a good solution actually, we just need to make sure that the old "loop" is not running when the "ready" event is fired by ioredis, so basically we should "cancel" previous loop proactively when starting the new one.

**Comment by tarunbatra** [comment]
Will pausing and resume the local queue ensure there are no competing "loops" running in the worker? 
```js
client.on('ready', async () => {
  if (reconnectingEvent) {
    reconnectingEvent = false;
    await queue.pause(true);
    await queue.resume(true);
  }
})
```
It appears to work for me at least in local.

**Comment by miwnwski** [comment]
@manast How would you approach this? Is there something we can check today (without monkey patching the lib) for an existing running loop?

**Comment by gabrieltomionardes** [comment]
I have the same issue using heroku, when it restart all queues get stuck in waiting.

**Comment by manast** [comment]
@miwnwski yeah it is a bit tricky, you need to somehow refactor the "run" method so that it can be cancellable https://github.com/OptimalBits/bull/blob/develop/lib/queue.js#L864
I could do it myself but I have a lot in my plate right now so it will take some time...

**Comment by anhnhatuit** [comment]
I have the same issue using GCP. Any quick solution about that.
Thank a lot for help.

**Comment by batrudinych** [comment]
Faced the same issue using AWS Elasticache. Looks like there was a connection interruption or some kind of a switch. No error logs or failed jobs, discovered the issue by missing data in the database.

**Comment by nirgilboa** [comment]
+1
Also encountering this issue when using AWS Elasticache

**Comment by dbousamra** [comment]
We are also encountering this issue.

**Comment by ilgianlu** [comment]
Hi, I'm working in a kubernetes cluster. Having the same issue. My workaround is to let the queue processor container die whenever it cannot connect to redis container. Kubernetes will take care of restarting my processor pod.
In order for the processor to exit abruptly I'm passing this redis conf:
const redisOptions = {
        ...config.get('redis.connectParams'),
        retryStrategy: (times) => {
            console.log('could not connect to redis!');
            process.exit(1);
        },
    };

**Comment by Xairoo** [comment]
Currently in dev and have the same problem. Appears after wake my system from hibernate. I am using a remote Redis server, so the connection gets dropped after a while when the system sleeps :-)

ioredis will reconnect automatically, but Bull won't pull new jobs.

**Comment by francois2metz** [comment]
Hi there. I do have some logs on our open source application https://github.com/1024pix/pix-editor

- At 2021-09-10T06:02:16.000Z, the connection to redis is dropped and the client is trying to reconnect until 2021-09-10T06:02:29.000Z,
- the connection is still not stable but commands have been resumed, but without the `brpoplpush` without means that the queues are not processed
- The connection is restarted every few minutes during days if I don't restart the application

Hope this helps? I'm available to debug anything

[extract_ioredis_logs.csv](https://github.com/OptimalBits/bull/files/7152385/extract_lcms_clean2.csv)

**Comment by Xairoo** [comment]
My MongoDB driver fires the disconnect event when I wake my system from sleep. It will directly reconnect.
ioredis doesn't trigger anything after sleep (but jobs get pushed to queue).

Hm, changed a while ago from Node Redis to ioredis. I think Node Redis hasn't this problem.

**Comment by Nyfy** [comment]
I am running into this issue as well when facing any kind of redis reconnection. This is very painful since the service appears healthy but in fact never processes any jobs. Any workaround available? Thanks.

**Comment by Xairoo** [comment]
Does this also happen to Bullmq?

**Comment by anhnhatuit** [comment]
> Hi, I'm working in a kubernetes cluster. Having the same issue. My workaround is to let the queue processor container die whenever it cannot connect to redis container. Kubernetes will take care of restarting my processor pod. In order for the processor to exit abruptly I'm passing this redis conf:
> 
> ```ts
> const redisOptions = {
>         ...config.get('redis.connectParams'),
>         retryStrategy: (times) => {
>             console.log('could not connect to redis!');
>             process.exit(1);
>         },
>     };
> ```

Sound good. I'm do manually to restart pod to make it works. I hope this is will help us resolve the issue.

Thanks,
Nhat

**Comment by manast** [comment]
@EPecherkin 
> It surprises me it wasn't fixed for a YEAR
Can you precise exactly what has not been fixed for a year?

**Comment by Nyfy** [comment]
@manast 

I can reproduce this issue by simply stopping and starting my Redis container. In my case I have a Node.JS app that acts as both a producer and consumer, and after stopping and starting my Redis container I can continue producing to the queues but I never receive callbacks to consume my repeated tasks.

The issue seems to have been well explained by @miwnwski above:

`I encountered this error and have been investigating, I'll share our findings here - maybe it helps. As far as I can understand, when the client reconnects, Queue#processJobs is not run. Queue.prototype.process calls Queue#start which calls Queue#run which calls Queue#processJobs which enters the queue in the forever loop state by throwing promises.

The problem (as I understand it) is that when Queue#processJobs is called it can resolve if the Queue is closing or if an error is thrown. This creates a state where the connection exists but the forever promise loop never executes.

Our patch fix for this (at least locally - not tested in production yet) is basically to call queue.run(queue.concurrency) on ready. Using the above example which I could use to reproduce the problem, this would be the "fix"`

I use the same workaround currently to get around this problem, but as you explained this is not a great solution:

`This sounds like a good solution actually, we just need to make sure that the old "loop" is not running when the "ready" event is fired by ioredis, so basically we should "cancel" previous loop proactively when starting the new one.

@miwnwski yeah it is a bit tricky, you need to somehow refactor the "run" method so that it can be cancellable https://github.com/OptimalBits/bull/blob/develop/lib/queue.js#L864
I could do it myself but I have a lot in my plate right now so it will take some time...`

It would be great if we could investigate this issue sooner rather than later since it seems to be affecting many production systems and is a silent failure in the worst way possible.

Thanks

**Comment by Xairoo** [comment]
@Nyfy how do you stop your redis container?
I get always `ECONNREFUSED` @ ioredis when stopping the container. And bull is working again when redis is back again.
Putting my system to sleep for about 10 minutes I will result in `ECONNRESET`.

Submitting new jobs will place them in redis, the worker needs some more time to get them: ~1 1/2 minutes. But my on complete handler doesn't receive the event, I had to restart the handler.

- API adds the job to the queue (I can see it in redis)
- Worker gets the job delayed
- Worker submits the job data when finished
- I can get `returnvalue` from redis
- ... but the on complete handler (a socket.io server) doesn't get the event

**Comment by manast** [comment]
@Nyfy If you can reproduce it consistently can you please provide me a code snippet that does? I also use docker to start and stop the redis container and I cannot reproduce it.

@Xairoo with esoteric I mean as a way of reproducing a software issue. We must find a way using software to reproduce the issue, should not be necessary to remove internet cables to do it.

**Comment by manast** [comment]
This is a simple code snippet you can try to play with to reproduce the issue:
```js
const Queue = require("bull");

const testQueue = new Queue("test", {
  redis: {
    maxRetriesPerRequest: null,
    enableReadyCheck: false,
  },
});

testQueue.process(async (job) => {
  console.log(`Process started for Job ${job.id}`);
});

testQueue.add(
  { foo: "bar" },
  {
    repeat: {
      every: 1500,
    },
  }
);

testQueue.on("error", (err) => console.log(err));

testQueue.on("global:completed", (jobId) => {
  console.log(`Job ${jobId} completed`);
});
```

I have tried extensively with ```docker star/stop/pause/unpause some-redis``` waiting different periods of time, but it always seems to be working.

**Comment by royalpinto** [comment]
> 🎉 This issue has been resolved in version 4.0.0 🎉
> 
> The release is available on:
> 
> * [npm package (@latest dist-tag)](https://www.npmjs.com/package/bull/v/4.0.0)
> * [GitHub release](https://github.com/OptimalBits/bull/releases/tag/v4.0.0)
> 
> Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot 📦🚀

@manast I can confirm it works for me for the environment given here https://github.com/luin/ioredis/issues/1285#issuecomment-903768612 when the ioredis gets reconnected.

**Comment by gramcha** [comment]
```
maxRetriesPerRequest: null,
enableReadyCheck: false,
```

setting this will not work ioredis 4+ version at the moment due to the following reason.

ioredis opens Redis connection in normal use cases using the StandaloneConnector implementation and that implementation only cares about port, host, family, path, and tls. Sadly ignores params like maxRetriesPerRequest, enableReadyCheck, and due that default value 20 and true assigned respectively.

Bull lib throwing error with below message when we try to use BULL 4.2.1.
```
Error: Using a redis instance with enableReadyCheck or maxRetriesPerRequest is not permitted.
  see https://github.com/OptimalBits/bull/issues/1873
```

Not sure how to proceed further, @manast

**Comment by Yuliang-Lee** [comment]
After dozen tests, I found the problem was cause by our redis server's deploy-way. LOL. Our redis server was run on k8s with original redis's config -- obviously there is no persistent data when redis server restarts. I did not have time to dig why redis's data lost cause the problem but it is no any problem happens when I modify redis data dir to persistence volumn.

**Comment by Shahtaj-Khalid** [comment]
I'm still facing the same issue, bull process stops consuming new messages upon reset connection error from redis :
created a separate issue with more details here : https://github.com/OptimalBits/bull/issues/2612.

setting these flags explicitly didn't help : 
```
maxRetriesPerRequest: null,
enableReadyCheck: false,
```

I'm using Bull's version : `4.10.2` & ioredis : `5.2.4`

Does anyone found a work around on how to fix this? or do I need to update to any other version of Bull? Kindly let me know as I'm really stuck, thanks.
